sdks/java/io/cassandra/src/test/java/org/apache/beam/sdk/io/cassandra/CassandraIOTest.java

```diff
@@ -0,0 +1,231 @@
+    final FakeCassandraService service = new FakeCassandraService();
+    service.load();
+
+    PipelineOptions pipelineOptions = PipelineOptionsFactory.create();
+    CassandraIO.CassandraSource source = new CassandraIO.CassandraSource(
+        null,
+        null,
+        new SerializableFunction<PipelineOptions, CassandraService>() {
+      @Override
+      public CassandraService apply(PipelineOptions input) {
+        return service;
+      }
+    });
+    long estimatedSizeBytes = source.getEstimatedSizeBytes(pipelineOptions);
+    assertEquals(estimatedSizeBytes, 115890);
+  }
+
+  @Test
+  @Category(ValidatesRunner.class)
+  public void testRead() throws Exception {
+    FakeCassandraService service = new FakeCassandraService();
+    service.load();
+
+    PCollection<Scientist> output = (PCollection<Scientist>) pipeline.apply(CassandraIO
+        .<Scientist>read()
+        .withCassandraService(service)
+        .withTable("scientist")
+        .withCoder(SerializableCoder.of(Scientist.class))
+        .withEntity(Scientist.class)
+    );
+
+    PAssert.thatSingleton(output.apply("Count", Count.<Scientist>globally()))
+        .isEqualTo(10000L);
+
+    PCollection<KV<String, Integer>> mapped =
+        output.apply(
+            MapElements.via(
+                new SimpleFunction<Scientist, KV<String, Integer>>() {
+                  public KV<String, Integer> apply(Scientist scientist) {
+                    KV<String, Integer> kv = KV.of(scientist.name, scientist.id);
+                    return kv;
+                  }
+                }));
+    PAssert.that(mapped.apply("Count occurrences per scientist", Count.<String, Integer>perKey()))
+        .satisfies(
+            new SerializableFunction<Iterable<KV<String, Long>>, Void>() {
+              @Override
+              public Void apply(Iterable<KV<String, Long>> input) {
+                for (KV<String, Long> element : input) {
+                  assertEquals(element.getKey(), 1000, element.getValue().longValue());
+                }
+                return null;
+              }
+            });
+
+    pipeline.run();
+  }
+
+  /**
+   * A {@link CassandraService} implementation that stores the entity in memory.
+   */
+  private static class FakeCassandraService implements CassandraService {
+
+    private static final Map<Integer, Scientist> table = new HashMap<>();
+
+    public void load() {
+      table.clear();
+      String[] scientists = {
+          "Einstein",
+          "Darwin",
+          "Copernicus",
+          "Pasteur",
+          "Curie",
+          "Faraday",
+          "Newton",
+          "Bohr",
+          "Galilei",
+          "Maxwell"
+      };
+      for (int i = 0; i < 10000; i++) {
+        int index = i % scientists.length;
+        Scientist scientist = new Scientist();
+        scientist.id = i;
+        scientist.name = scientists[index];
+        put(scientist);
+      }
+    }
+
+    public void put(Scientist scientist) {
```

Comment:
this method is unnecessary - it's used only at line 146, just inline. Not sure if get() is the same way.